cobalt/browser/metrics/cobalt_metrics_logs_uploader.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright 2025 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/browser/metrics/cobalt_metrics_logs_uploader.h"
+
+#include "base/base64url.h"
+#include "base/logging.h"
+#include "cobalt/browser/h5vcc_metrics/public/mojom/h5vcc_metrics.mojom.h"
+#include "components/metrics/log_decoder.h"
+#include "components/metrics/metrics_log_uploader.h"
+#include "mojo/public/cpp/bindings/remote.h"
+#include "third_party/metrics_proto/chrome_user_metrics_extension.pb.h"
+#include "third_party/metrics_proto/cobalt_uma_event.pb.h"
+#include "third_party/metrics_proto/reporting_info.pb.h"
+
+namespace cobalt {
+
+using browser::metrics::CobaltUMAEvent;
+
+// Helper method to create a trimmed down and sanitized version of the UMA
+// proto for Cobalt
+void PopulateCobaltUmaEvent(
+    const ::metrics::ChromeUserMetricsExtension& uma_proto,
+    const ::metrics::ReportingInfo reporting_info_proto,
+    CobaltUMAEvent& cobalt_proto) {
+  cobalt_proto.mutable_histogram_event()->CopyFrom(uma_proto.histogram_event());
+  cobalt_proto.mutable_user_action_event()->CopyFrom(
+      uma_proto.user_action_event());
+  cobalt_proto.mutable_reporting_info()->CopyFrom(reporting_info_proto);
+}
+
+void CobaltMetricsLogUploader::UploadLog(
+    const std::string& compressed_log_data,
+    const std::string& log_hash,
+    const std::string& log_signature,
+    const metrics::ReportingInfo& reporting_info) {
+  // Always run the upload complete callback as a success. Arguments to callback
+  // don't matter much here as we're not really doing anything but forwarding to
+  // the H5vcc API.
+  base::ScopedClosureRunner closure(
+      base::BindOnce(on_upload_complete_, /*status*/ 200, /* error_code */ 0,
+                     /*was_https*/ true, false, ""));
+
+  if (!metrics_listener_.is_bound()) {
+    return;
+  }
+
+  if (service_type_ == ::metrics::MetricsLogUploader::UMA) {
```

Comment:
nit: 
```
  if service_type != ...) {
    return;
  }
```
bc prefer early return.

Reply:
I commented on this elsewhere. In this case an early return doesn't make sense as this function can handle many different types of services.

Reply:
But the code right now doesn't handle those other services, so I'd expect either:
```
  if (service_type_ == A) {
    // do stuff
  }
  // TODO(): Handle other |service_type_| types?
  // Maybe NOTIMPLEMENTED() << "Only A is supported);
  return OK;
```
Or
```
  if (service_type_ != A) {
    return something;
  } 
  // Do stuff.
  return something;
```

The current code is neither, and as a general rule we should prefer return early and avoid nesting. Hence the review comment.

Reply:
Happy to make this change... but where are you getting this "general rule" from? I see no mention of early return in neither the Chromium or Google C++ style guide. Also, generally, this is the first time it's ever been mentioned to me as a best practice.

Reply:
It might have been in some early "Chromium Dos and Donts", but now I can' t find it. I'm 100% sure this was told to me in the past in multiple reviews though : )

I found a discussion in chromium-dev about it ([here](https://groups.google.com/a/chromium.org/g/chromium-dev/c/NO2CUzFVmUA)). I'm pretty sure this is the recommended style (a Google search has quite a few hits explaining why it's better to nesting loops). 

Happy to propose a CL to fix https://source.chromium.org/chromium/chromium/src/+/main:styleguide/c++/c++-dos-and-donts.md upstream and see where the chips fall ?

Reply:
So that chromium-dev thread basically amounts to: "use whatever is best for your situation" (which, I agree with). Up to you if you want to upstream the do/dont thing. This feels very arbitrary and like something that doesn't really have a performance win as I assume the compiler would optimize such things. In this specific case, it's not really enhancing readability either as it's a single "if" and also semantically more correct for the intent of the function.